Add Open XL toolchain and config changes for z/OS

Deigue opened the pull request:
Adds the initial set of changes to support compilation with Open XL on z/OS. This will define the alternative flags and configuration that needs to be used by the wyvern compiler while running on z/OS.

(This is one part of the multiple changes added for supporting Open XL compilation on OMR)

babsingh commented:
jenkins build all

Deigue commented:
Changes here confirmed passing [j11](https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK11_s390x_zos_Personal/4114/console) and [j21](https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK21_s390x_zos_Personal/566/) builds.

babsingh commented:
jenkins build all

babsingh reviewed (commented):
High level feedback:
- Squash the two commits into a single commit.
- Revise and improve comments.
- Remove dead code.

babsingh commented:
@keithc-ca Can you please review this PR?

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,261 @@
+	set(OMR_C_WARNINGS_AS_ERROR_FLAG -qhalt=w)
+	set(OMR_CXX_WARNINGS_AS_ERROR_FLAG -qhalt=w)
+
+	# There is no enhanced warning for XLC right now
```
This file should not be used for xlc: why is that compiler even mentioned here?

keithc-ca commented on ddr/Makefile:
```diff
@@ -26,7 +26,7 @@ include $(top_srcdir)/omrmakefiles/configure.mk
 OMR_TOOLCHAIN ?= gcc
 
-ifneq (,$(filter gcc xlc,$(OMR_TOOLCHAIN)))
+ifneq (,$(filter gcc xlc openxl,$(OMR_TOOLCHAIN)))
```
Do we care about non-cmake builds with OpenXL? If not, the remaining changes should be removed.

keithc-ca commented:
The commit message and the discussion here should refer to "z/OS", not "zos".

Deigue commented on cmake/modules/OmrDetectSystemInformation.cmake:
```diff
@@ -202,12 +202,18 @@ macro(omr_detect_system_information)
+				set(CMAKE_C_COMPILER_IS_OPENXL ON CACHE BOOL "ibm-clang is the C compiler")
+				set(_OMR_TOOLCONFIG "openxl")
+			else()
+				message(STATUS "NO MATCH")
```
Hi Keith, I was verifying all my cleanup changes with different builds. After making the above mentioned changes, while Open XL passes with no problems, XLC build gives a fatal error, because I guess it is entering thiis z/OS section, but not matching with the other xlclang compilers and in place of the "NO MATCH" its just throwing the error.

Perhaps it makes more sense to update this to 
`message(STATUS "NO XLClang match. Using xlc toolchain.")`

Does this sound good?

Deigue commented:
Thanks for the feedback/cleanup suggestions. I've addressed the above comments and verified builds are passing.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,232 @@
+		# For XLClang, remove any usages of -qhalt=e or -qhalt=s provided by default
+		# in the CMAKE CXX/C/ASM FLAGS, since xlclang/xlclang++ are not compatible
+		# with the e or s options.
+		omr_remove_flags(CMAKE_ASM_FLAGS -qhalt=e)
+		omr_remove_flags(CMAKE_CXX_FLAGS -qhalt=s)
+		omr_remove_flags(CMAKE_C_FLAGS   -qhalt=e)
```
Suggest languages should be addressed in a consistent order; see lines 34-35 and elsewhere, where C is handled before C++.

Deigue replied:
reordered to put C_FLAGS first as seen in other places.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,232 @@
+			-m64
+		)
+	else()
+		# -march should be there for 32 and 64 C/CXX flags but the C compiler is used for
+		# the assembler and it has trouble with some assembly files if it is specified.
```
Why does this comment reference 64-bit compiles here in the 32-bit branch?

Deigue replied:
removed along with the rest of "ppc" section, as POWER support will be contributed by another developer as mentioned below.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,232 @@
+	endmacro(omr_toolconfig_global_setup)
+endif()
+
+if(OMR_HOST_ARCH STREQUAL "ppc")
```
Has any of this been tested on POWER? I see several `-q` options which appear to be in the xlc style.

Deigue replied:
Another dev (Jackie) is working on support for POWER, as such I will be removing this untested section from here. Believe she has commits that are adding support separately so this section should be not here.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,232 @@
+		set(CMAKE_C_ARCHIVE_CREATE "<CMAKE_AR> -X64 cr <TARGET> <LINK_FLAGS> <OBJECTS>")
+		set(CMAKE_C_ARCHIVE_FINISH "<CMAKE_RANLIB> -X64 <TARGET>")
+	endif()
+
```
Please omit this blank line.

Deigue replied:
Removed extra line.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,232 @@
+
+	set(CMAKE_ASM_FLAGS "-fno-integrated-as")
+	string(APPEND CMAKE_ASM_FLAGS " \"-Wa,-mgoff\"")
+	string(APPEND CMAKE_ASM_FLAGS " \"-Wa,-mSYSPARM(BIT64)\"")
```
Shouldn't `BIT64` be conditional on `OMR_ENV_DATA64`?

Deigue replied:
Added the missing conditional

keithc-ca commented on configure:
```diff
@@ -1727,7 +1727,8 @@ Some influential environment variables:
   OMR_TOOLCHAIN
-              The toolchain used to build the package. One of: gcc,xlc,msvc
+              The toolchain used to build the package. One of:
+              gcc,xlc,msvc,openxl
```
This file should not be modified: it's not used in cmake builds.

Deigue replied:
removed

keithc-ca commented on configure.ac:
```diff
@@ -47,7 +47,7 @@ AC_ARG_VAR([exe_output_dir], [The output directory for executables. (Default: $t
 AC_ARG_VAR([OMR_HOST_ARCH], [The architecture of the CPU where the package will run. One of: aarch64,arm,ppc,riscv,s390,x86])
 AC_ARG_VAR([OMR_TARGET_DATASIZE], [Specifies whether the package will run in 32- or 64-bit mode. One of: 31,32,64])
-AC_ARG_VAR([OMR_TOOLCHAIN], [The toolchain used to build the package. One of: gcc,xlc,msvc])
+AC_ARG_VAR([OMR_TOOLCHAIN], [The toolchain used to build the package. One of: gcc,xlc,msvc,openxl])
```
This file should not be modified: it's not used in cmake builds.

Deigue replied:
removed

Deigue commented:
Should have addressed all the comments above, also changed to use `-std=gnu++14` as recommended by compiler team earlier. Added change here as it belongs part of the toolchain files commits.

Deigue commented:
Added conditional `elseif(CMAKE_C_COMPILER_ID MATCHES "IBMClang$")`  as on Jenkins builds the compiler ID is detected as `IBMClang 2.1.0.0` , but locally it may resolve as "zOS"

Deigue commented:
@keithc-ca Addressed above comments and ready for review. Awaiting approval on this from your side, then will reach out to Babneet for reviewing again and merging.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,180 @@
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+###############################################################################
+
+if(CMAKE_C_COMPILER_IS_XLCLANG)
+	macro(omr_toolconfig_global_setup)
+		# For XLClang, remove any usages of -qhalt=e or -qhalt=s provided by default
+		# in the CMAKE CXX/C/ASM FLAGS, since xlclang/xlclang++ are not compatible
+		# with the e or s options.
+		omr_remove_flags(CMAKE_ASM_FLAGS -qhalt=e)
+		omr_remove_flags(CMAKE_C_FLAGS   -qhalt=e)
+		omr_remove_flags(CMAKE_CXX_FLAGS -qhalt=s)
+	endmacro(omr_toolconfig_global_setup)
+endif()
```
This file should not be used when `CMAKE_C_COMPILER_IS_XLCLANG` (in which case `OMR_TOOLCONFIG` is `xlc`). All the resulting dead code should be removed from this file.

Even after such changes this would still use `-q` options, which I believe are for xlc, not OpenXL.

Deigue replied:
Thanks, removed these from the toolcfg file.

keithc-ca commented on cmake/modules/OmrDetectSystemInformation.cmake:
```diff
@@ -198,19 +198,28 @@ macro(omr_detect_system_information)
 		elseif(CMAKE_C_COMPILER_ID MATCHES "^XL(Clang)?$" OR CMAKE_C_COMPILER_ID STREQUAL "zOS")
 			# In CMake 3.14 and prior, XLClang uses CMAKE_C_COMPILER_ID "XL"
 			# In CMake 3.15 and beyond, XLClang uses CMAKE_C_COMPILER_ID "XLClang"
 			set(_OMR_TOOLCONFIG "xlc")
```
This belongs within the `if` block for xlc.

Deigue replied:
Just to confirm, does this mean the check would be updated as such:
```
elseif(CMAKE_C_COMPILER_ID MATCHES "^XL(Clang)?$" OR  CMAKE_C_COMPILER_ID MATCHES "IBMClang$" OR CMAKE_C_COMPILER_ID STREQUAL "zOS") 
```

I only had it as a separate clause because it could not enter the `elseif` mentioned above, and the COMPILER_ID was ending up to be `IBMClang` when running with Open XL via a jenkins build.

Deigue replied:
Have updated rest of feedback @keithc-ca , just awaiting a bit of clarification for the above change.

keithc-ca replied:
`CMAKE_C_COMPILER_IS_OPENXL` should not be define, but there are references elsewhere so it will need to stay temporarily.

I was suggesting that line 205 should be
```
		elseif((CMAKE_C_COMPILER_ID MATCHES "IBMClang$") OR (CMAKE_C_COMPILER MATCHES ".*ibm-clang.*"))
```
and lines 219-221 should be removed.

Deigue replied:
oh I see, that is a lot cleaner. I misunderstood that suggestion ... fixing it up.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,169 @@
+###############################################################################
+
+if(OMR_OS_AIX)
+	list(APPEND OMR_PLATFORM_C_COMPILE_OPTIONS -qlanglvl=extended)
+	list(APPEND OMR_PLATFORM_CXX_COMPILE_OPTIONS -qlanglvl=extended0x)
```
These look like xlc options, not openxl options.

Deigue replied:
@midronij Which files are currently being used by AIX for Open XL? Looks like I have this section specific to AIX where the flags aren't updated. I am assuming if the changes for AIX are already merged, they may be coming from somewhere else already. Is it safe to assume this section can be removed, just wanted to confirm.

midronij replied:
On AIX, OpenXL actually uses the GNU cmake file (`cmake/modules/platform/toolcfg/gnu.cmake`). Generally speaking, clang-based compilers (like OpenXL or xlclang) are considered to be GNU compilers by the config and make files. 

That being the case, I opted to modify the existing GNU cmake files to account for the use of OpenXL on AIX, rather than create new ones for OpenXL. So you should be able to safely remove that section and any others that are specific to AIX. The only thing to be careful of is that the addition of this new file doesn't break the AIX build, since my changes don't take its existence into account.

Deigue replied:
yes, it should be mutually exclusive then, Thanks. I will remove the above section.

keithc-ca commented on cmake/modules/ddr/DDRSetStub.cmake.in:
```diff
@@ -145,7 +145,7 @@ function(process_source_files src_files)
 			list(APPEND pp_command "-xc++")
 		endif()
-		if("@OMR_OS_ZOS@" AND "@OMR_ENV_DATA64@")
+		if("@OMR_OS_ZOS@" AND "@OMR_ENV_DATA64@" AND NOT "@CMAKE_C_COMPILER_IS_OPENXL@")
 			# We need to set 64 bit code generation to ensure that limits.h macros are set correctly.
 			list(APPEND pp_command "-Wc,lp64")
```
Does openxl only support 64-bit targets?

Deigue replied:
Don't think so, but this exclusion was simply added so that the build knows not append this flag. For Open-XL, this is applied by default without specifying. Adding it to the full command with the "-Wc" format causes compilation error as it is not recognized.

keithc-ca replied:
I think `CMAKE_C_COMPILER_IS_OPENXL` should not be used, in favour of using `OMR_TOOLCONFIG STREQUAL "openxl"`.

Deigue commented:
Made the requested changes based on Keith's feedback and also rebased with latest `master`.
Please review.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,145 @@
+
+	list(APPEND OMR_PLATFORM_CXX_COMPILE_OPTIONS
+		-march=${OMR_ZOS_COMPILE_ARCHITECTURE}
+		"-std=gnu++14"
```
Why `gnu` and not `c` as in `-std=c++14`?

Deigue replied:
Sean from the cpp team recommended to use `gnu++14` over `c++14` to avoid using `_EXT` in most places ...
More specifics regarding feedback there:
`_EXT` required for alloca() not being exposed in `javadump.cpp`
```
Use -std=gnu++14.  The closest XL option to -std=c++14 is -qlanglvl=ansi.  With that XL option you would need to add 
_EXT too.  The -std=gnu* options are the closest to -qlanglvl=etended.   stdlib.h will declare alloca() with -std=gnu*.
---
If you say you want a standard language level (like std=c++14), then you do need to specify the _EXT to say you want
extensions. If you use extension based language levels (like the gnu versions) then you get most extensions enabled by
default.
```

keithc-ca replied:
Ok. Please add a comment that summarizes that.

keithc-ca replied:
I don't see that comment.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,145 @@
+		-mnocsect
+	)
+
+	# Configure the platform dependent library for multithreading.
```
Please hyphenate "platform-dependent".

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,145 @@
+
+set(SPP_CMD ${CMAKE_C_COMPILER})
+
+if(CMAKE_C_COMPILER_IS_XLCLANG)
+	# The -P option doesn't sit well with XLClang, so it's not included. It causes:
+	# "ld: 0711-317 ERROR: Undefined symbol: <SYMBOL>" when libj9jit29.so is getting linked.
+	set(SPP_FLAGS -E)
+else()
+	set(SPP_FLAGS -E -P)
+endif()
```
Isn't `CMAKE_C_COMPILER_IS_XLCLANG` only true when `OMR_TOOLCONFIG` is `xlc` (in which case this file would not be used)?

Deigue replied:
yes, will remove the conditional and keep the path thats always taken for open-xl.

Deigue commented:
Addressed comments above. please review.

Deigue commented:
Addressed comments above, please review @keithc-ca

keithc-ca commented on cmake/modules/ddr/DDRSetStub.cmake.in:
```diff
@@ -145,7 +145,7 @@ function(process_source_files src_files)
 			list(APPEND pp_command "-xc++")
 		endif()
-		if("@OMR_OS_ZOS@" AND "@OMR_ENV_DATA64@")
+		if("@OMR_OS_ZOS@" AND "@OMR_ENV_DATA64@" AND NOT OMR_TOOLCONFIG STREQUAL "openxl")
```
`OMR_TOOLCONFIG` must be referenced like other variables used in this file:
```
		if("@OMR_OS_ZOS@" AND "@OMR_ENV_DATA64@" AND NOT "@OMR_TOOLCONFIG@" STREQUAL "openxl")
```

Deigue commented:
@keithc-ca Have made the above suggested changes. Apologies for the missing comment, I had accidentally added it somewhere else and it did not make it into earlier commit. Should be there now.

keithc-ca commented on cmake/modules/platform/toolcfg/openxl.cmake:
```diff
@@ -0,0 +1,140 @@
+		-march=${OMR_ZOS_COMPILE_ARCHITECTURE}
+	)
+
+	# -std=gnu++14 is used over -std=cpp++14, as the gnu* options are closest to 
+	# -qlanglvl=extended (Enabling most extensions by default). This is preferred
+	# in order to avoid defining _EXT repeatedly in most places.
```
Please remove the trailing space on line 46; "enabling" should not be capitalized.

babsingh commented:
jenkins build all

babsingh commented:
Infra failure on riscv, but this PR's changes don't impact riscv. So, we can merge it without a riscv PR build.
